### PR TITLE
Incorporate definitions for Masonry

### PR DESCRIPTION
**Pull request by m-a-wilson**
case 1. Add a new type definition.
- [x] checked compilation succeeds with `--target es6` and `--noImplicitAny` options.
- [x] has correct [naming convention](http://definitelytyped.org/guides/contributing.html#naming-the-file)
- [x] has a [test file](http://definitelytyped.org/guides/contributing.html#tests) with the suffix of  `-tests.ts` or `-tests.tsx`.

### CODE REVIEW

File: masonry-layout/masonry-layout.d.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+declare module Masonry {
+
+     class Masonry implements MasonryGrid { 
+        constructor(options?: MasonryOptions); 
+        constructor(selector: string, options?: MasonryOptions); 
+    } 
+
+     interface MasonryGrid {
+        masonry?(): void;
```
Comment (mhegazy) [inline masonry-layout/masonry-layout.d.ts]: why is this optional?

**Comment by m-a-wilson** [comment]
By 'this' I assume you are referring to `masonry?(): void;`. As I noted in the comments, I simply modified the original to pass the checks and respond to the requests from the reviewer.

Normally Masonry is invoked with some number of options, usually at least two:
```
$('.grid').masonry({
  columnWidth: 200,
  itemSelector: '.grid-item'
});
```
It will work without any options set though the result is not really all that nice:
 ```
$('.grid').masonry();
```
I'm open to any better way to express that.

**Comment by mhegazy** [comment]
The question mark here indicates that the property itself is optional. i.e. `$('.grid').masonary` may exist or not. in other words, it can be undefined. i do not think this is the intent. 

As for the optionality of the arguments, you have added two overload signatures, that allows for this, i.e.:

```ts
masonry(): void;
masonry(eventName: string, listener: any): void;
```

which indicates that you can call `masonry` with no arguments or with two.

you can find more information about this in http://www.typescriptlang.org/docs/handbook/functions.html#overloads and http://www.typescriptlang.org/docs/handbook/interfaces.html#optional-properties

**Comment by m-a-wilson** [comment]
Looks like my initial approach (assume the basic file was correct and only had minor structural issues) was incorrect. Several options use 'any' unnecessarily. The class definition is using 'implements' incorrectly. I'll update the corrected files as soon as I finish testing.

**Comment by gjuchault** [comment]
Could you also add Vanilla typings ?

**Comment by m-a-wilson** [comment]
I'll add it to the list.

**Comment by gjuchault** [comment]
I've come up with something like that (vanilla only):

```ts
declare module 'masonry' {
  type Elements = Element | NodeList | Array<Element>;

  export default class Masonry {
    constructor(element: string | Element, options: MasonryOptions);

    layout(): void;
    layoutItems(items: Array<any>, isStill: boolean);
    stamp(elements: Elements);
    unstamp(elements: Elements);
    appended(elements: Elements);
    prepended(elements: Elements);
    addItems(elements: Elements);
    remove(elements: Elements);
    on(eventName: string, listener: Function);
    off(eventName: string, listener: Function);
    once(eventName: string, listener: Function);
    reloadItems(): void;
    destroy(): void;
    getItemsElement(): Array<Element>;

    static data(element: string | Element): Masonry;
  }

  export interface ContainerStyles {
    [key: string]: string;
  }

  export interface MasonryOptions {
    columnWidth: number | string | Element;
    itemSelector: string;
    percentPosition?: boolean;
    gutter?: number | string | Element;
    stamp?: string;
    fitWidth?: boolean;
    originLeft?: boolean;
    isOriginLeft?: boolean;
    originTop?: boolean;
    isOriginTop?: boolean;
    containerStyle?: ContainerStyles,
    transitionDuration?: number | string;
    stagger?: number | string;
    resize?: boolean;
    isResizeBound?: boolean;
    initLayout?: boolean;
    isInitLayout?: boolean;
  }
}
```

I've only one `any` for layoutItems as I don't know what's a `Masonry.Item` (from the docs [here](http://masonry.desandro.com/methods.html#layoutitems))

**Comment by paulvanbrenk** [comment]
@m-a-wilson gentle ping

**Comment by jamesond** [comment]
Is this still a work in progress to use the typescript.

**Comment by m-a-wilson** [comment]
The PR has been merged. I’m still looking into suggested changes for the vanilla version. I have no pride of ownership here. If someone else can make it better, go for it.

**Pull request merged**